core/src/memory.rs

```diff
@@ -376,15 +379,26 @@ impl AllocationsContext {
         Ok(start)
     }
 
-    /// Free specific page.
-    ///
-    /// Currently running program should own this page.
+    /// Free specific memory page.
     pub fn free(&mut self, page: WasmPage) -> Result<(), AllocError> {
-        if page < self.static_pages || page >= self.max_pages || !self.allocations.remove(&page) {
-            Err(AllocError::InvalidFree(page.0))
-        } else {
-            Ok(())
+        self.free_range(page..=page)
+    }
+
+    /// Try to free pages in range. Will only return error if range is invalid.
+    ///
+    /// Currently running program should own this pages.
+    pub fn free_range(&mut self, range: RangeInclusive<WasmPage>) -> Result<(), AllocError> {
+        if *range.start() < self.static_pages || *range.end() >= self.max_pages {
+            let page = if *range.start() < self.static_pages {
```

Comment:
excess comparison for no reason

Reply:
it is not exess because comparison above is __OR__,

Reply:
This's still **IS**.

Following snippet does the same with less amount of lines, being more readable and a very little bit more optimised, isn't it?

```rust
if *range.start() < self.static_pages {
    return Err(AllocError::InvalidFree(range.start().0));
}

if *range.end() >= max_pages {
    return Err(AllocError::InvalidFree(range.end().0));
}
```

Reply:
I've forgot I've introduced new error for this, take a look